working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+thread plus some associated runtime structures (e.g. thread local allocation
+buffer). Domains are _not_ isolated from each other: they allocate objects in a
+global heap, which is shared between all domains and can access and mutate
+shared global state.
+
+[PEP 703]: (https://peps.python.org/pep-0703/)
+[communicating event-loops]: http://www.erights.org/talks/promises/paper/tgc05.pdf
+[transferrable objects]: https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects
+[`SharedArrayBuffer`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer
+[Using Web Workers]: https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers
+[Ruby Thread]: https://docs.ruby-lang.org/en/3.2/Thread.html
+[`threading`]: https://docs.python.org/3/library/threading.html
+[Fearless Concurrency]: https://doc.rust-lang.org/book/ch16-00-concurrency.html#fearless-concurrency
+[reference capabilities]: https://bluishcoder.co.nz/2017/07/31/reference_capabilities_consume_recover_in_pony.html
+[Joe Armstrong PhD Thesis]: https://erlang.org/download/armstrong_thesis_2003.pdf
+[Swift concurrency]: https://docs.swift.org/swift-book/documentation/the-swift-programming-language/concurrency/
+[`Sendable` protocol]: https://github.com/apple/swift-evolution/blob/main/proposals/0302-concurrent-value-and-concurrent-closures.md
+[communicating sequential processes]: https://www.cs.cmu.edu/~crary/819-f09/Hoare78.pdf
+[go-concurrency]: https://go.dev/tour/concurrency/1
+[effective-go-sharing]: https://go.dev/doc/effective_go#sharing
+[multicore-ocaml-timeline]: https://tarides.com/blog/2023-03-02-the-journey-to-ocaml-multicore-bringing-big-ideas-to-life/
+[Retrofitting Parallelism onto OCaml]: https://core.ac.uk/download/pdf/328720849.pdf
+[Retrofitting Effect Handlers onto OCaml]: https://kcsrk.info/papers/drafts/retro-concurrency.pdf
+[ocaml-domain]: https://v2.ocaml.org/manual/parallelism.html#s:par_domains
+
+## Shareable Data
+
+I propose to extend Dart with a shared memory multithreading but provide a clear
+type-based delineation between two concurrency worlds. **Only instances of
+classes implementing `Shareable` interface can be concurrently mutated by
+another thread.** We restrict the kind of data that a shareable class can
+contain: **fields of shareable classes can only contain references to instances
+of shareable classes.** This requirement is enforced at compile time by
+requiring that static types of all fields are shareable.
+
+```dart
+// dart:core
+
+/// [Shareable] instances can be shared between isolates in
+/// a group and mutated concurrently.
+///
+/// A class implementing [Shareable] can only declare fields
+/// which have a static type that is a subtype of [Shareable].
+abstract interface class Shareable {
+}
+```
+
+```dart
+class S implements Shareable {
+  // ...
+}
+```
+
+> [!NOTE]
+>
+> I choose marker interface rather than dedicated syntax (e.g.
```

Comment:
Nit: I -> We

Reply:
I have chosen to consistently write `I` everywhere to mark _my opinion_ as the author of the proposal (e.g. `I propose` etc) and `we` everywhere to refer to Dart SDK developers (e.g. `we will implement`).